Profiling workflows

kazcw opened the pull request:
### Pull Request Description

Define some workflows for batch-mode profiling.

Implemented:
- collapse nodes
- create node
- enter collapsed node
- new project
- open visualization

They can currently be built and run with a command like:
`./run.sh ide build --profiling-level=debug && dist/ide/linux-unpacked/enso --entry-point profile --workflow create_node --save-profile out.json`

And the data can be displayed with:
`dist/ide/linux-unpacked/enso --entry-point profiling_run_graph --load-profile out.json`

Demo of recording and viewing a profile with a command-line one-liner:

https://user-images.githubusercontent.com/1047859/169954795-2d9520ca-84f9-45d2-b83a-5063ebe6f718.mp4

See: https://www.pivotaltracker.com/story/show/182195399.

### Important Notes

- When defining workflows, two helpers are enough to allow us to tell when the action is really done: `Fixture::compile_new_shaders`, and `Fixture::backend_execution`. Often, it is appropriate to await both, but it depends on the task.
- The shader compiler is now driven by a `Controller`; while the `Compiler` is reset if context is lost, the `Controller`'s state survives context loss.
- A new `--load-profile` option supports specifying a profile by path when running `profiling_run_graph`.
- Drop the `with_same_start` profiler interface; we ended up preferring a child profiler convention, and this interface was not implemented compatibly with the stricter data model we've had since the introduction of `profiler::data`.
- Fix the noisy `rustfmt` output.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run.sh ide dist` and `./run.sh ide watch`.

[ci no changelog needed]

kazcw committed:
Remove dead code. `remote` is obsolete.

kazcw committed:
Fix rustfmt warnings

kazcw committed:
Profiling workflows

kazcw committed:
Merge branch 'develop' into wip/kw/profiling-workflows

kazcw committed:
lint and docs

kazcw committed:
Merge branch 'develop' into wip/kw/profiling-workflows

wdanilo commented on app/gui/view/graph-editor/src/lib.rs:
```diff
@@ -3901,3 +3917,98 @@ mod tests {
+    // === Node creation ===
+    // =====================
+
+    // FIXME: This functionality overlaps with the [`tests::add_node_by`] functions above.
```
Fix left in the code

kazcw replied:
The duplication was already present, I just added the note. I can't refactor it right now because I can't get the integration tests to work. I opened a bug (https://www.pivotaltracker.com/story/show/182278293) and added a link in the comment.

wdanilo commented on lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs:
```diff
@@ -71,14 +74,16 @@ pub struct Job<T> {
     #[deref]
     input:    T,
     handler:  WeakJobHandler,
+    profiler: profiler::Debug,
 }
 
 impl<T> Job<T> {
-    fn with_mod_input<S>(self, f: impl FnOnce(T) -> S) -> Job<S> {
+    fn map_input<S>(self, f: impl FnOnce(T) -> S) -> Job<S> {
```
Oh wow, the old name was bad indeed :D

wdanilo commented on lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs:
```diff
@@ -229,21 +239,22 @@ impl CompilerData {
     fn run(&mut self, time: animation::TimeInfo) -> bool {
         let mut any_new_shaders_ready = false;
         if self.dirty {
+            let _profiler = profiler::start_debug!(profiler::APP_LIFETIME, "shader::Compiler::run");
```
can we extract that to a function and use the function annotation instead? It's way nicer than this.

wdanilo commented on lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs:
```diff
@@ -485,3 +499,75 @@ impl Error {
+// === Controller ===
+// ==================
+
+/// Controls the shader compiler. While the shader compiler will be thrown away if context is lost,
+/// the controller's state is persistent.
+#[derive(Debug, Clone, CloneRef, Default)]
+pub struct Controller {
+    rc: Rc<RefCell<ControllerData>>,
+}
+
+#[derive(Default, Derivative)]
```
1. The docs doesnt explain what is the purpose this struct exists. I suspect its needed for performance tracking, but I know it only because it is part of this PR, I would never know it when reading the code - this should be explained in the docs.
2. Could you use our callback manager instead here? There are a lot of different callback registries defined, like `callback::registry::CopyMut1` - I believe you'd find something with exactly this functionality there. However, if you came to a conclusion that no implementation there has matching performance to this one, I'd love to know why, because these callback registries were meant to be lightweight and used widely.

wdanilo reviewed (approved):
1. Amazing to see it! :) 
2. Add docs to the readme about how to run it. Take core of making the docs clear and easy to follow, please.
3. A few comments left in the code. I'd love someone else to make the review as well, as I was doing it in a bit of rush today.

kazcw committed:
Introduce `processes` tool

New tool can generate CSV reports about how much time each process
(frontend, shader compilation, backend) takes to complete each profiling
workflow.

kazcw committed:
wip csv

kazcw committed:
refactor and docs

kazcw committed:
Revert "wip csv"

This reverts commit 4fb3670d45eb4672ffc8bfe60802d3c697d12ab4.

kazcw committed:
add node about integration test failures

kazcw committed:
Merge branch 'develop' into wip/kw/profiling-workflows

kazcw committed:
Merge branch 'develop' into wip/kw/profiling-workflows

kazcw committed:
lint

kazcw commented:
Since this still needs more review anyway, I'm going to merge in the changes for the tool that generates the parallel process latency reports (frontend vs shaders vs backend)--it's a small changeset, and it depends on this one.

It's run like this:
```
ubuntu@enso:~/repo/enso/wip/lib/rust/profiler/data$ cargo run --bin processes compile_new_shaders,backend_execution < profile.json
```

It produces CSV data that can be graphed like this:
![chart](https://user-images.githubusercontent.com/1047859/170763219-8a3d787f-18aa-475e-8f58-30c69882d31f.png)

The chart shows, for different workflows, which process (frontend/backend/shaders) is holding up completion of the workflow; as you can see, backend is usually much slower. The bottlenecking on the backend will be more stark after some planned performance improvements: The current text area implementation is super inefficient, and slated for replacement. After it is fixed, time spent compiling shaders in response to user interactions should be eliminated, and frontend time should be roughly halved, depending on workflow.

kazcw committed:
Add workflow needed for tools in upcoming branch build-faster

MichaelMauderer commented on app/gui/src/integration_test.rs:
```diff
@@ -104,6 +120,88 @@ impl Drop for Fixture {
+    /// Open a project; doesn't complete until the project is ready to render.
+    #[profile(Objective)]
+    pub async fn open_project() -> Self {
+        let template_project = "Orders";
```
This could probably have a better name and comment indicating that "Orders" is the default project used.

MichaelMauderer commented on app/gui/src/lib.rs:
```diff
@@ -142,23 +142,26 @@ mod profile_workflow;
+        "debug_mode_is_active",
+        analytics::AnonymousData(debug_mode),
+    );
+    let config =
```
Why is this inlined instead of the previous `start_and_forget` method?

kazcw replied:
`start_and_forget` did not just `start()` and forget; it had logic for hiding the loading indicator, which is not applicable to any of the other users of `Initializer`.

MichaelMauderer commented on app/gui/src/model/module.rs:
```diff
@@ -301,6 +302,29 @@ pub struct Notification {
+    }
+
+    /// Return a profiler to be attached to a new [`Notification`].
+    fn profiler() -> profiler::Debug {
```
Should this functionality be part of the `profiler::internal` module? At first glance, I do not see anything that would need to be done in this place of the source.

kazcw replied:
Implemented an interface for creating paused profilers.

MichaelMauderer commented on app/gui/view/graph-editor/src/lib.rs:
```diff
@@ -3901,3 +3917,100 @@ mod tests {
+/// Utilities for simulating user interaction, to support testing and profiling.
+// This is not #[cfg(test)] because it is needed for batch-mode profiling,
+// which is not just in testing builds.
+pub mod automation {
```
Could this be a separate file?

MichaelMauderer commented on app/ide-desktop/lib/client/src/index.js:
```diff
@@ -547,6 +553,17 @@ function createWindow() {
+    const profiles = Promise.all(profilePromises)
+    Electron.ipcMain.on('load-profiles', event => {
+        profiles.then(profiles => {
+            const n = profiles.length
```
`n` doesn't seem to be used anywhere.

MichaelMauderer commented on lib/rust/profiler/data/src/bin/processes.rs:
```diff
@@ -0,0 +1,268 @@
+    measurement_process: &mut HashMap<MeasurementId, Process>,
+    profile: &Profile<OpaqueMetadata>,
+    process_by_label: &HashMap<String, Process>,
+    m: MeasurementId,
```
`m` could be named better: `measurement_id`

MichaelMauderer commented on app/gui/src/profile_workflow.rs:
```diff
@@ -40,20 +43,43 @@ pub async fn entry_point_profile() {
-    assert_eq!(graph_editor.nodes().all.len(), initial_nodes + added_nodes);
+async fn profile_create_node() {
+    let test = Fixture::create_project().await;
+    let _profiler = profiler::start_objective!(profiler::APP_LIFETIME, "@highlight");
```
Can `@highlight` be moved to some constant?

kazcw replied:
Profiler labels must be literals to enable packing file/line info into the string, which has a large impact on runtime profiling data size.

kazcw committed:
review

kazcw committed:
review

kazcw committed:
Merge remote-tracking branch 'origin/develop' into profiling-workflows

kazcw committed:
size limit
